AI/BattleAI/BattleAI.cpp

```diff
@@ -552,12 +556,139 @@ void CBattleAI::print(const std::string &text) const
 	logAi->trace("%s Battle AI[%p]: %s", playerID.getStr(), this, text);
 }
 
+bool CBattleAI::isItOurLastTurn(PossibleSpellcast::ValueMap & values, const std::vector<battle::Units> & queue, HypotheticBattle * state)
+{
+	bool firstRound = true;
+	bool ourFirstTurn = true;
+	bool enemyHadTurn = false;
+
+	for(auto & round : queue)
+	{
+		if (!firstRound)
+		{
+			state->nextRound(0);//todo: set actual value?
+			firstRound = false;
+		}
+		for(auto unit : round)
+		{
+			LOGL("---------------> consider turn of " + unit->unitType()->namePl);
+			if (!state->getForUpdate(unit->unitId())->alive())
+			{
+				LOGL("It will be dead...");
+				if (unit->alive())
+				{
+					LOGL("but it's alive according to simple method");
+				}
+				else
+				{
+					LOGL("It's also dead according to simple method");
+				}
+				
+				continue;
+			}
+
+			if(state->battleGetOwner(unit) == playerID)
+			{
+				if (!ourFirstTurn)
+				{
+					LOGL("Our unit will have turn!");
+					return false;
+				}
+				else 
+				{
+					LOGL("this is this turn!");
+				}
+				ourFirstTurn = false;
+			}
+			else
+			{
+				LOGL("Enemy will have turn!");
+				enemyHadTurn = true;
+			}
+			state->nextTurn(unit->unitId());
+
+			PotentialTargets pt(unit, state);
+
+			LOGL("I predict:");
+			if(!pt.possibleAttacks.empty())
+			{
+				AttackPossibility ap = pt.bestAction();
+
+				auto swb = state->getForUpdate(unit->unitId());
+				*swb = *ap.attackerState;
+
+				if(ap.damageDealt > 0)
+					swb->removeUnitBonus(Bonus::UntilAttack);
+				if(ap.damageReceived > 0)
+					swb->removeUnitBonus(Bonus::UntilBeingAttacked);
+
+				LOGL(unit->unitType()->namePl + " will attack:");
+				for(auto affected : ap.affectedUnits)
+				{
+					LOGL("->"+affected->unitType()->namePl);
+					swb = state->getForUpdate(affected->unitId());
+					*swb = *affected;

```

Comment:
There is more generic problem with algorithm (same for spells): after this line queue is actually invalid and should be rebuilt. `battleGetTurnOrder` should be called for each unit and from `state`